### PR TITLE
Add TikTokUserIE.

### LINKED ISSUE #674: [Site Request] Tiktok Profiles
## Checklist

- [x] I'm reporting a site feature request
- [x] I've verified that I'm running yt-dlp version **2021.08.10**
- [x] I've searched the bugtracker for similar site feature requests including closed ones

## Description

For Tiktok when I attempt to download an individual video using urls such as https://www.tiktok.com/@meme/video/6994875996952448261 it successfully downloads the Tiktoks but it would be nice to be able to just put the profile url and download all the Tiktoks on the account.

### PR DESCRIPTION
**Pull request by Ashish0804**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Needs cookies to work.
Closes https://github.com/yt-dlp/yt-dlp/issues/674

### CODE REVIEW

File: yt_dlp/extractor/tiktok.py [hunk 1]
```diff
@@ -150,3 +152,52 @@ def _real_extract(self, url):
...
+        'info_dict': {
+            'id': '79005827461758976',
+        },
+        'skip': 'Cookies are required to download from TikTok user account.'
+    }]
+
+    def _real_extract(self, url):
+        id = self._match_id(url)
+        webpage = self._download_webpage(url, id)
+        user_id = re.search(r'\"id\":\"(?P<userid>\d+)', webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:179]: _search_regex

File: yt_dlp/extractor/tiktok.py [hunk 2]
```diff
@@ -150,3 +152,52 @@ def _real_extract(self, url):
...
+        if not user_id:
+            raise ExtractorError('Cookies are required to download from TikTok user account.', expected=True)
+        user_id = user_id.group('userid')
+        secuid = re.search(r'\"secUid\":\"(?P<secUid>[^\"]+)', webpage).group('secUid')
+        verifyFp_cookie = self._get_cookies('https://www.tiktok.com').get('s_v_web_id')
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:184]: `verifyfp_cookie`

File: yt_dlp/extractor/tiktok.py [hunk 3]
```diff
@@ -150,3 +152,52 @@ def _real_extract(self, url):
...
+        if not verifyFp_cookie:
+            raise ExtractorError('Improper cookies (missing s_v_web_id).', expected=True)
+        api_url = f'https://m.tiktok.com/api/post/item_list/?aid=1988&cookie_enabled=true&count=30&verifyFp={verifyFp_cookie.value}&secUid={secuid}&cursor='
+        hasMore = True
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:188]: snake case

File: yt_dlp/extractor/tiktok.py [hunk 4]
```diff
@@ -150,3 +152,52 @@ def _real_extract(self, url):
...
+        cursor = '0'
+        entries = []
+        page = 0
+        while hasMore:
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:192]: itertools.count

File: yt_dlp/extractor/tiktok.py [hunk 5]
```diff
@@ -150,3 +152,52 @@ def _real_extract(self, url):
...
+            data_json = self._download_json(api_url + cursor, user_id, note='Downloading Page %d' % page)
+            cursor = data_json['cursor']
+            hasMore = data_json['hasMore']
+            videos = data_json.get('itemList', [])
+            for video in videos:
+                video_id = video['id']
+                entries.append(self.url_result('https://www.tiktok.com/@%s/video/%s' % (id, video_id),
+                                               ie=TikTokIE.ie_key(),
+                                               video_id=video_id))
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:201]:
```suggestion
                entries.append(self.url_result('https://www.tiktok.com/@%s/video/%s' % (id, video_id),
                                               ie=TikTokIE.ie_key(), video_id=video_id))
```

File: yt_dlp/extractor/tiktok.py [hunk 6]
```diff
@@ -150,3 +152,52 @@ def _real_extract(self, url):
...
+        },
+        'skip': 'Cookies are required to download from TikTok user account.'
+    }]
+
+    def _real_extract(self, url):
+        id = self._match_id(url)
+        webpage = self._download_webpage(url, id)
+        user_id = re.search(r'\"id\":\"(?P<userid>\d+)', webpage)
+        if not user_id:
+            raise ExtractorError('Cookies are required to download from TikTok user account.', expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:181]: raise_login_required

File: yt_dlp/extractor/tiktok.py (hunk 4 above)
Comment (Ashish0804) [reply yt_dlp/extractor/tiktok.py:192]: the while loop is seen as 1 iteration then.
To overcome this, will have to put all funcs in the for iter loop and then break when has_more is False.
Doesn't look readable imo.

File: yt_dlp/extractor/tiktok.py (hunk 6 above)
Comment (Ashish0804) [reply yt_dlp/extractor/tiktok.py]: Changed the message.
raise_login_required is not necessary now.

File: yt_dlp/extractor/tiktok.py [hunk 7]
```diff
@@ -150,3 +150,50 @@ def _real_extract(self, url):
...
+        while has_more:
+            data_json = self._download_json(api_url + cursor, user_id, note='Downloading Page %d' % page)
+            cursor = data_json['cursor']
+            has_more = data_json['hasMore']
+            videos = data_json.get('itemList', [])
+            for video in videos:
+                video_id = video['id']
+                entries.append(self.url_result('https://www.tiktok.com/@%s/video/%s' % (id, video_id),
+                                               ie=TikTokIE.ie_key(), video_id=video_id))
+            page += 1
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:198]: extract lazily

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tiktok.py [hunk 8]
```diff
@@ -127,3 +128,79 @@ def _real_extract(self, url):
...
+                    'duration': try_get(video, lambda x: x['video']['duration'], int),
+                    'view_count': try_get(video, lambda x: x['stats']['playCount'], int),
+                    'like_count': try_get(video, lambda x: x['stats']['diggCount'], int),
+                    'comment_count': try_get(video, lambda x: x['stats']['commentCount'], int),
+                    'repost_count': try_get(video, lambda x: x['stats']['shareCount'], int),
+                    'timestamp': video.get('createTime'),
+                    'creator': try_get(video, lambda x: x['author']['nickname'], str),
+                    'uploader': try_get(video, lambda x: x['author']['uniqueId'], str),
+                    'uploader_id': try_get(video, lambda x: x['author']['id'], str),
+                    'uploader_url': 'https://www.tiktok.com/@' + user_id,
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:192]:
```suggestion
                    'uploader_url': f'https://www.tiktok.com/@{user_id}',
```

File: yt_dlp/extractor/tiktok.py [hunk 9]
```diff
@@ -127,3 +128,79 @@ def _real_extract(self, url):
...
+        webpage = self._download_webpage(url, user_id)
+        own_id = self._search_regex(r'\"id\":\"(?P<userid>\d+)', webpage, user_id, default=None)
+        if not own_id:
+            raise ExtractorError('Cookies (not necessarily logged in) are needed.', expected=True)
+        secuid = self._search_regex(r'\"secUid\":\"(?P<secUid>[^\"]+)', webpage, user_id)
+        verifyfp_cookie = self._get_cookies('https://www.tiktok.com').get('s_v_web_id')
+        if not verifyfp_cookie:
+            raise ExtractorError('Improper cookies (missing s_v_web_id).', expected=True)
+        api_url = f'https://m.tiktok.com/api/post/item_list/?aid=1988&cookie_enabled=true&count=30&verifyFp={verifyfp_cookie.value}&secUid={secuid}&cursor='
+        has_more = True
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:162]:
```suggestion
```

**Pull request merged**